src/probability_theory/probability_space.lean

```diff
@@ -0,0 +1,777 @@
+  apply le_trans _ h2,
+  apply nnreal_le_sub_add',
+end
+
+lemma em_event_cond {Ω:Type*} {p:probability_space Ω} (A B:event Ω): (A ∩ B ∪ A ∩ Bᶜ) = A :=
+begin
+  apply event.eq,
+  simp [set.inter_union_compl],
+end
+
+lemma disjoint_inter_compl {α:Type*} (A B C:set α):disjoint (A ∩ B) (C∩ Bᶜ) :=
+begin
+  apply set.disjoint_of_subset_left (set.inter_subset_right A B),
+  apply set.disjoint_of_subset_right (set.inter_subset_right C Bᶜ),
+  simp [disjoint_iff],
+end
+
+lemma Pr_em {Ω:Type*} {p:probability_space Ω} (A B:event Ω):
+  Pr[A ∩ B] + Pr[A ∩ Bᶜ] = Pr[A] :=
+begin
+  rw ← Pr_disjoint_eor,
+  rw em_event_cond,
+  simp [disjoint_inter_compl],
+end
+
+lemma Pr_diff {Ω:Type*} {p:probability_space Ω} (A B:event Ω):
+    A.val ⊆ B.val →
+    Pr[B ∩ Aᶜ] = Pr[B] - Pr[A] :=
+begin
+  intro A1,
+  have A2:Pr[B ∩ A] + Pr[B ∩ Aᶜ] = Pr[B],
+    apply Pr_em,
+  have A3:(B ∩ A) = A,
+    apply eand_eq_self_of_subset_right,
+    apply A1,
+  rw A3 at A2,
+  symmetry,
+  apply nnreal_add_sub_left,
+  exact A2,
+end
+
+/- Define the equivalance of two events. -/
+def event_eqv {Ω:Type*} {p:probability_space Ω} (A B:event Ω):event Ω := A ∩ B ∪ Aᶜ ∩ Bᶜ
+
+infixr `=ₑ`:100 := event_eqv
+
+lemma event_eqv_def {Ω:Type*} {p:probability_space Ω} (A B:event Ω):
+    (A =ₑ B) = ((A ∩ B) ∪ ((Aᶜ) ∩ (Bᶜ))) := rfl
+
+lemma eor_partition {Ω:Type*} {p:probability_space Ω} (A B:event Ω):
+  (A ∪ B) = ((A ∩ Bᶜ) ∪ (A ∩ B) ∪ (Aᶜ ∩ B)) :=
+begin
+  apply event.eq,
+  simp,
+  ext ω,split;intros A1;simp at A1;simp [A1],
+    cases A1 with A1 A1; simp [A1],
+    rw or_comm,
+    apply classical.em,
+    apply classical.em,
+    cases A1 with A1 A1,
+      tauto,
+    cases A1 with A1 A1,
+    simp [A1],
+end
+
+lemma Pr_eor_partition {Ω:Type*} {p:probability_space Ω} (A B:event Ω):
+  Pr[A ∪ B] = Pr[A ∩ Bᶜ] + Pr[A ∩ B] + Pr[Aᶜ ∩ B] :=
+begin
+  rw eor_partition A B,
+  rw Pr_disjoint_eor,
+  rw Pr_disjoint_eor,
+  simp,
+  rw set.disjoint_left,
+  intros ω A1,
+  simp at A1,
+  simp [A1],
+  simp,
+  split;
+  {rw set.disjoint_left,
+  intros ω A1,
+  simp at A1,
+  simp [A1]},
+end
+
+lemma Pr_eor_plus_eand {Ω:Type*}  {p:probability_space Ω} (A B:event Ω):
+  Pr[A ∪ B] + Pr[A ∩ B] = Pr[A] + Pr[B] :=
+begin
+  rw ← Pr_em A B,
+  rw ← Pr_em B A,
+  rw eand_comm B A,
+  rw eand_comm B (Aᶜ),
+  rw Pr_eor_partition A B,
+  ring,
+end
+
+lemma Pr_eor_eq_minus_eand {Ω:Type*}  {p:probability_space Ω} (A B:event Ω):
+  Pr[A ∪ B] = Pr[A] + Pr[B] - Pr[A ∩ B] :=
+begin
+  rw ← Pr_eor_plus_eand,
+  rw nnreal.add_sub_cancel,
+end
+
+lemma Pr_eor_eq_minus_eand_real {Ω:Type*}  {p:probability_space Ω} (A B:event Ω):
+  (Pr[A ∪ B]:real) = (Pr[A]:real) + (Pr[B]:real) - (Pr[A ∩ B]:real) :=
+begin
+  have A1:Pr[A ∪ B] + Pr[A ∩ B] = (Pr[A] + Pr[B]),
+  {apply Pr_eor_plus_eand},
+  rw ← nnreal.coe_eq at A1,
+  repeat {rw nnreal.coe_add at A1},
+  linarith,
+end
+
+def eall_encodable {Ω β:Type*} {p:probability_space Ω} [encodable β] (A:β → event Ω):event Ω :=
```

Comment:
This is `infi`

Reply:
We don't have a complete_lattice instance

Reply:
Perhaps we can define `Inf` as `if h : encodable β then /- this -/ else ⊥`